### PR TITLE
Fix random ray solver to correctly simulate fixed source problems with fissionable materials

### PR DESCRIPTION
**Pull request by jtramm**
# Description

Currently the flat source random ray solver does not correctly add in fission source contributions when in fixed source mode. The linear source solver works correctly already, so no change is needed there.

This PR fixes the flat source calculation to add in the fission contribution even when in fixed source mode. I also added a new test for fixed source subcritical multiplication to test this condition. The pytest problem doesn't fully converge things in the interest of runtime, but the well converged results are given below so as to validate that the solver works for this sort of problem.

# Validation

The test problem is a variant on the 2x2 example lattice typically used in testing for the random ray mode. The case removes two of the pincells, such that the 2x2 lattice has three fully moderator cells and one fueled pin. There is also a small source placed in the corner of one of the moderator regions. The true k-eff of the system is around 0.84, so it can be safely simulated with MGMC in fixed source mode so as to provide a reference solution.

### Source Region Mesh:

Screenshot 2024-08-01 at 10 15 29 AM

### Fixed Source location:

Screenshot 2024-08-01 at 10 15 12 AM

### Example flux solution:

Screenshot 2024-08-01 at 10 16 08 AM

### Results:

Tallies are defined by the flux in each group, with a 2x2 spatial mesh (resulting in 28 tally bins). Accuracy is compared against MGMC using the same 7 group cross sections. As can be seen below, the error for the current develop is quite high (90%) for most tallies. However, with the changes to the flat source treatment in this PR the flat source error drops down to around 1 or 2%. As mentioned earlier, linear source was already working, so we can see that the error for that run is around 0.13% on average.

Screenshot 2024-08-02 at 3 54 51 PM

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: src/random_ray/flat_source_domain.cpp
Comment (yardasol) [inline src/random_ray/flat_source_domain.cpp:161]: So it looks like Flat Source random ray wasn't pulling fission cross sections _at all_ for eigenvalue calculations? Great catch, glad it's a simple fix!

**Review by yardasol — APPROVED** [review]
Nice catch @jtramm!

**Comment by yardasol** [comment]
I'll wait for @paulromano's stamp of approval before merging since you requested him as a reviewer.

**Comment by paulromano** [comment]
Thanks @yardasol, and sorry @jtramm for the delay on this. I'll take a look shortly!

File: src/random_ray/flat_source_domain.cpp
Comment (jtramm) [reply src/random_ray/flat_source_domain.cpp:161]: For eigenvalue calculations it was including the fission cross sections as expected, but for fixed source w/flat sources, it was indeed omitting any fission contributions.

Comment (yardasol) [reply src/random_ray/flat_source_domain.cpp:161]: Ah, I miswrote, yeah that's what I meant to say.

**Review by paulromano — APPROVED** [review]
Thanks @jtramm!

**Pull request merged**
